### DISCUSSION TITLE
My thoughts

### DISCUSSION THREAD

**Discussion by adleong** [linkerd/polixy#1] [General]
Generally this seems like a very nice and straightforward way to implement access control.  Here are the thoughts I had when reading over this spec and the associated examples.  These range from broad questions about future work in this space to the very nit picky.

> Instead, the discovered configuration indicates the IPs on which connections are permitted, and the proxy only forwards connections targeting these IPs.

We might need to update our proxy-init rules to accommodate this (there may be assumptions that all inbound traffic not addressed to localhost should go to the proxy)... not sure.

> Use of these headers may be disabled by setting a server annotation:

Seems a bit unusual to have this be an annotation on the server resource rather than a field inside the server resource spec.

> "Wyld Stallyns problem"

Excellent.

> HTTP route authorization

The possibility of adding this in the future is exciting.  Having different authorization policies for different routes or gRPC functions seems very cool.  There's a lot to think about regarding how this will converge with ServiceProfiles.

> References servers in the same namespace by name or label selector.

Is it necessary for a ServerAuthorization to be able to select Servers by selector rather than just directly by name?  For comparison, RoleBindings can only select Roles by name.  I see the example in `proxy-admin.yml` where a ServerAuthorization is shared between two different Servers, but it doesn't seem unreasonable to me to require defining a ServerAuthorization for each.

The `meshTLS` schema is defined as a `oneOf`.  Should it be a `anyOf` instead so that a mix of `identities`, and `serviceAccounts` can be provided?

Similarly, should the `clients` property accept an array so that multiple clients can be defined in a single ServerAuthorization.  For example, allowing unencrypted access from a specific subnet and requiring meshTLS for everything else.

**Comment by olix0r** [comment]
@adleong 

> > Instead, the discovered configuration indicates the IPs on which connections are permitted, and the proxy only forwards connections targeting these IPs.

> We might need to update our proxy-init rules to accommodate this (there may be assumptions that all inbound traffic not addressed to localhost should go to the proxy)... not sure.

Yeah. I think the proposed behavior should work >99% of the time, and hopefully we can flush out the other requirements with the change.

> > Use of these headers may be disabled by setting a server annotation:

> Seems a bit unusual to have this be an annotation on the server resource rather than a field inside the server resource spec.

Yeah, I agree. I _don't_ think this belongs on the `Server` resource, though, It belongs more properly as HTTP-specific configuration, on an HTTP-specific resource. Until we have proper HTTPRoute/GRPCRoute types, the annotation can fill the need. I can clarify this in the spec.

> Is it necessary for a ServerAuthorization to be able to select Servers by selector rather than just directly by name?

It's probably not strictly necessary. The one example I've used this with is for [admin servers](https://github.com/linkerd/polixy/blob/32c102f3fe3b1c26cd2e6f29d43fe34cf70887fb/k8s/linkerd/proxy-admin.yml#L41), of which there may be several, and it's convenient to have a blanket authorization for, for instance , prometheus. It seems like a nice shortcut to grant access to classes of server like this, but I could be convinced to take it out of the first version. Though, it doesn't make the implementation _that much_ worse, since we already have to watch servers...

> The meshTLS schema is defined as a oneOf. Should it be a anyOf instead so that a mix of identities, and serviceAccounts can be provided?

I think of the type as polymorphic, the authorization is permitting a class of traffic. The oneOf expresses this intent; however, I think implementation-wise it doesn't actually prevent what you're asking for. AIUI we're expressing that constraint that only one of the `required: [..]` constraints applies. But the required constraint does not enforce that none of the other fields are present!

> Similarly, should the clients property accept an array so that multiple clients can be defined in a single ServerAuthorization. For example, allowing unencrypted access from a specific subnet and requiring meshTLS for everything else.

Similarly as above, I don't think we're precluding expressing multiple client constraints, but I want to generally encourage using smaller targeted authorizations so that blast-radius of authorization changes is limited.

**Comment by mateiidavid** [comment]
I have a question around the `networks` field in the `ServerAuthz` crd and felt maybe it's better to just invade this conversation rather than create a separate discussion about it.

In [ServerAuthorization](https://github.com/linkerd/polixy/blob/main/DESIGN.md#serverauthorization) it is mentioned that:

> (the policy is) Scoped to source IP networks. If no networks are specified, the authorization applies to clients in all networks

What would happen if the network is not valid. Can there be an invalid network/subnet specified?  An obvious example that stands out to me is `255.255.255.255`, I'm just curious of whether we'd have to validate the policy in the controller or if we offload this to the end user? Might be an implementation detail so sorry about this, was just curious.
